File: src/main/java/pia/rest/ImageService.java
package pia.rest;

import kotlin.text.StringsKt;
import org.eclipse.jetty.util.StringUtil;
import org.glassfish.jersey.media.multipart.FormDataParam;
import pia.database.Database;
import pia.database.DatabaseQuery;
import pia.database.model.archive.Image;
import pia.exceptions.CreateHashException;
import pia.filesystem.BufferedFile;
import pia.logic.ImageReader;
import pia.logic.ImageWriter;
import pia.rest.contract.ErrorContract;
import pia.rest.contract.ImageApiContract;
import pia.rest.contract.ObjectList;
import pia.rest.contract.SuccessContract;
import pia.tools.FileHash;

import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Optional;
import java.util.UUID;

@Path("images")
public class ImageService {
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response queryImages() {
        ObjectList<ImageApiContract> imageApiContractObjectList = new ObjectList<>();
        DatabaseQuery<Image> query = Database.getConnection().query(Image.class);
        for(Image image : query.getAll()) {
            ImageApiContract contract = ImageApiContract.Companion.fromDb(image);
            imageApiContractObjectList.add(contract);
        }
        return Response.ok().entity(imageApiContractObjectList).build();
    }

    @Path("checkImageExistanceByHash")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response checkImageExistanceByHash(@QueryParam("hashType") String hashType,
                                              @QueryParam("hash") String hash) {
        Response response;
        String hashTypeInternal = hashType;
        if(StringUtil.isEmpty(hashTypeInternal)) {
            hashTypeInternal = "SHA-256";
        }
        ImageReader reader = new ImageReader();
        if(reader.findImagesBySHA256(hash).size() > 0) {
            response = Response.status(Response.Status.OK).build();
        } else {
            response = Response.status(Response.Status.NO_CONTENT).build();
        }
        return response;
    }

    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    public Response addImage(@FormDataParam("image")InputStream imageStream,
                             @FormDataParam("fileName")String fileName) {
        Response response = null;
        try {
            ImageReader reader = new ImageReader();
            BufferedFile bufferedFile = new BufferedFile(imageStream.readAllBytes());
            Optional<String> hash = FileHash.createHash(bufferedFile);
            if(hash.isPresent()) {
                if(reader.findImagesBySHA256(hash.get()).size() == 0) {
                    ImageWriter writer = new ImageWriter();
                    writer.addImage(bufferedFile, fileName);
                } else {
                    response = Response.status(Response.Status.FOUND).entity(new ErrorContract("imagefile already uploaded")).build();
                }
            } else {
                response = Response.serverError().entity(new ErrorContract("could not create hash for image")).build();
            }
        } catch (IOException e) {
            response = Response.serverError().entity(new ErrorContract(e)).build();
        } catch (CreateHashException e) {
            response = Response.serverError().entity(new ErrorContract(e)).build();
        } catch (Throwable e) {
            response = Response.serverError().entity(new ErrorContract(e)).build();
        }
        if(response == null) {
            response = Response.accepted().build();
        }
        return response;
    }

    @Path("{imageId}/getFile")
    @GET
    @Produces(MediaType.APPLICATION_OCTET_STREAM)
    public Response getImageFile(@PathParam("imageId")String imageId) {
        UUID imaageUUID = UUID.fromString(imageId);
        ImageReader imageReader = new ImageReader();
        Optional<InputStream> fileStream = imageReader.getImageFileByImageIdFromDisk(imaageUUID);

        Response response;
        if(fileStream.isPresent()) {
            response = Response.ok().entity(fileStream.get()).build();
        } else {
            response = Response.noContent().build();
        }
        return response;
    }

    @Path("{imageId}")
    @DELETE
    @Produces(MediaType.APPLICATION_JSON)
    public Response deleteImage(@PathParam("imageId")String imageId) {
        Response response;
        UUID imageUUID = UUID.fromString(imageId);
        ImageReader imageReader = new ImageReader();
        Optional<Image> image = imageReader.findImageById(imageUUID);
        if(image.isPresent()) {
            ImageWriter imageWriter = new ImageWriter();
            if(imageWriter.deleteImage(image.get())) {
                response = Response.ok().entity(new SuccessContract("successfully deleted")).build();
            } else {
                response = Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(new ErrorContract("error deleting the image")).build();
            }
        } else {
            response = Response.status(Response.Status.NOT_FOUND).build();
        }
        return response;
    }

    @Path("deleteAllDebug")
    @DELETE
    @Produces(MediaType.APPLICATION_JSON)
    public Response deleteAllDebug() {
        ImageWriter writer = new ImageWriter();
        DatabaseQuery<Image> query = Database.getConnection().query(Image.class);
        for(Image image : query.getAll()) {
            writer.deleteImage(image);
        }
        return Response.ok().build();
    }
}

File: src/main/java/pia/rest/ImageService.kt
package pia.rest

import jetbrains.exodus.query.GetAll
import jetbrains.exodus.query.NodeBase
import kotlinx.dnq.query.eq
import kotlinx.dnq.query.query
import kotlinx.dnq.query.toList
import mu.KotlinLogging
import org.eclipse.jetty.util.StringUtil
import org.glassfish.jersey.media.multipart.FormDataParam
import pia.database.Database
import pia.database.model.archive.Image
import pia.database.model.archive.StagedType
import pia.exceptions.InternalException
import pia.logic.FileStager
import pia.logic.ImageReader
import pia.logic.ImageWriter
import pia.logic.WorkBalancer
import pia.rest.contract.ErrorContract
import pia.rest.contract.ImageApiContract
import pia.rest.contract.ImageApiContract.Companion.fromDb
import pia.rest.contract.ObjectList
import pia.rest.contract.SuccessContract
import pia.tools.toEntityId
import java.io.InputStream
import java.net.URI
import java.time.LocalDateTime
import java.util.*
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("images")
class ImageService {
    val logger = KotlinLogging.logger { }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun queryImages(@QueryParam("missingFilePath") missingFilePath: Boolean): Response {
        var response: Response
        response = try {
            val imageApiContractObjectList = ObjectList<ImageApiContract>()
            Database.connection.transactional(true) {

                for (image in Image.query(applyFilter(missingFilePath)).toList()) {
                    val contract = fromDb(image)
                    imageApiContractObjectList.add(contract)
                }
            }
            Response.ok().entity(imageApiContractObjectList).build()
        } catch (e: java.lang.Exception) {
            logger.error(e) { }
            Response.serverError().build()
        }
        return response
    }

    private fun applyFilter(missingFilePath: Boolean): NodeBase {
        var nodeBase: NodeBase? = null
        if (missingFilePath) {
            nodeBase = Image::pathToFileOnDisk eq ""
        }
        if (nodeBase == null) {
            nodeBase = GetAll()
        }
        return nodeBase
    }

    @Path("checkImageExistanceByHash")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun checkImageExistanceByHash(
        @QueryParam("hashType") hashType: String?,
        @QueryParam("hash") hash: String?
    ): Response {
        var response: Response
        try {
            var hashTypeInternal = hashType
            if (StringUtil.isEmpty(hashTypeInternal)) {
                hashTypeInternal = "SHA-256"
            }
            val reader = ImageReader()
            response = if (reader.findImagesBySHA256(hash!!).size > 0) {
                Response.status(Response.Status.OK).build()
            } else {
                Response.status(Response.Status.NO_CONTENT).build()
            }
        } catch (e: java.lang.Exception) {
            logger.error(e) { }
            response = Response.serverError().build()
        }
        return response
    }

    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    fun addImage(
        @FormDataParam("image") imageStream: InputStream,
        @FormDataParam("fileName") fileName: String?,
        @FormDataParam("creationTimeStamp") creationTimeStamp: LocalDateTime?
    ): Response {
        //logger.info("CurrentRunningUploads: " + CurrentRunningUploadCounter.currentRunningUploads)
        WorkBalancer.startStagingFile()
        val response: Response = try {
            imageStream.use {
                FileStager().stageFile(imageStream, fileName.orEmpty(), StagedType.Image)
                Response.created(URI.create("")).build()
            }
        } catch (e: Throwable) {
            logger.error(String.format("error adding image %s", fileName), e);
            Response.serverError().entity(ErrorContract(e)).build()
        }
        WorkBalancer.endStagingFile()
        return response
    }

    @Path("{imageId}/getFile")
    @GET
    @Produces(MediaType.APPLICATION_OCTET_STREAM)
    fun getImageFile(@PathParam("imageId") imageId: String): Response {
        val response: Response = try {
            val imageReader = ImageReader()
            val fileStream = imageReader.getImageFileByImageIdFromDisk(imageId.toEntityId())
            if (fileStream.isPresent) {
                Response.ok().entity(fileStream.get()).build()
            } else {
                Response.noContent().build()
            }
        } catch (e: Exception) {
            logger.error(e) { }
            Response.serverError().build()
        }
        return response
    }

    @Path("{imageId}")
    @DELETE
    @Produces(MediaType.APPLICATION_JSON)
    fun deleteImage(@PathParam("imageId") imageId: String): Response {
        var response: Response
        try {
            val imageUUID = UUID.fromString(imageId)
            val imageReader = ImageReader()
            val image = imageReader.findImageById(imageId.toEntityId())
            response = if (image.isPresent) {
                val imageWriter = ImageWriter()
                if (imageWriter.deleteImage(image.get(), false)) {
                    Response.ok().entity(SuccessContract("successfully deleted")).build()
                } else {
                    Response.status(Response.Status.INTERNAL_SERVER_ERROR)
                        .entity(ErrorContract("error deleting the image")).build()
                }
            } else {
                Response.status(Response.Status.NOT_FOUND).build()
            }
        } catch (e: java.lang.Exception) {
            logger.error(e) { }
            response = Response.serverError().build()
        }
        return response
    }

    @Path("deleteAllDebug")
    @DELETE
    @Produces(MediaType.APPLICATION_JSON)
    fun deleteAllDebug(): Response {
        var response: Response
        try {
            val writer = ImageWriter()
            Database.connection.transactional {
                for (image in Image.all().toList()) {
                    if (!writer.deleteImage(image, true)) {
                        throw InternalException("unable to delete image ${image.pathToFileOnDisk}")
                    }
                }
                it.commit()
            }
            response = Response.ok().build()
        } catch (e: java.lang.Exception) {
            logger.error(e) {}
            response = Response.serverError().build()
        }
        return response
    }
}